python/instrumentation/openinference-instrumentation-langchain/src/openinference/instrumentation/langchain/_tracer.py

```diff
@@ -0,0 +1,440 @@
+        "inputs": (INPUT_VALUE, INPUT_MIME_TYPE),
+        "outputs": (OUTPUT_VALUE, OUTPUT_MIME_TYPE),
+    }.items():
+        span.set_attributes(dict(zip(io_attributes, _convert_io(run.get(io_key)))))
+    span.set_attributes(
+        dict(
+            _flatten(
+                chain(
+                    _prompts(run["inputs"]),
+                    _input_messages(run["inputs"]),
+                    _output_messages(run["outputs"]),
+                    _prompt_template(run["serialized"]),
+                    _invocation_parameters(run),
+                    _model_name(run["extra"]),
+                    _token_counts(run["outputs"]),
+                    _function_calls(run["outputs"]),
+                    _tools(run),
+                    _retrieval_documents(run),
+                )
+            )
+        )
+    )
+
+
+def _flatten(key_values: Iterable[Tuple[str, Any]]) -> Iterator[Tuple[str, AttributeValue]]:
+    for key, value in key_values:
+        if value is None:
+            continue
+        if isinstance(value, Mapping):
+            for sub_key, sub_value in _flatten(value.items()):
+                yield f"{key}.{sub_key}", sub_value
+        elif isinstance(value, List) and any(isinstance(item, Mapping) for item in value):
+            for index, sub_mapping in enumerate(value):
+                for sub_key, sub_value in _flatten(sub_mapping.items()):
+                    yield f"{key}.{index}.{sub_key}", sub_value
+        else:
+            if isinstance(value, Enum):
+                value = value.value
+            yield key, value
+
+
+Message = Dict[str, Any]
+
+
+def _langchain_run_type_to_span_kind(run_type: str) -> OpenInferenceSpanKindValues:
+    try:
+        return OpenInferenceSpanKindValues(run_type.upper())
+    except ValueError:
+        return OpenInferenceSpanKindValues.UNKNOWN
+
+
+def _serialize_json(obj: Any) -> str:
+    if isinstance(obj, datetime):
+        return obj.isoformat()
+    return str(obj)
+
+
+def _convert_io(obj: Optional[Dict[str, Any]]) -> Iterator[str]:
+    if not obj:
+        return
+    if not isinstance(obj, dict):
+        raise ValueError(f"obj should be dict, but obj={obj}")
+    if len(obj) == 1 and isinstance(value := next(iter(obj.values())), str):
+        yield value
+    else:
+        yield json.dumps(obj, default=_serialize_json)
+        yield OpenInferenceMimeTypeValues.JSON.value
+
+
+def _prompts(run_inputs: Dict[str, Any]) -> Iterator[Tuple[str, List[str]]]:
+    """Yields prompts if present."""
+    if "prompts" in run_inputs:
+        yield LLM_PROMPTS, run_inputs["prompts"]
+
+
+def _input_messages(run_inputs: Mapping[str, Any]) -> Iterator[Tuple[str, List[Message]]]:
+    """Yields chat messages if present."""
+    if not hasattr(run_inputs, "get"):
+        return
+    # There may be more than one set of messages. We'll use just the first set.
+    if not (multiple_messages := run_inputs.get("messages")):
+        return
+    assert isinstance(
```

Comment:
I see a vulnerability here in that our parsing logic raises assertion errors when the data is of an unexpected format, which will crash `end_trace`. It looks like that will prevent a span from being created. More importantly, is the user's application insulated from that crash, meaning even if an assertion error is raised here, will the user's app at least continue to function with no traces?

Reply:
added try statements